File: library/src/main/java/com/thesurix/gesturerecycler/LayoutFlags.java
package com.thesurix.gesturerecycler;


import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.support.v7.widget.helper.ItemTouchHelper;

/**
 * Enum with predefined gesture flags for various layout managers, see {@link RecyclerView.LayoutManager}
 * @author thesurix
 */
enum LayoutFlags {

    LINEAR {
        @Override
        int getDragFlags(final RecyclerView.LayoutManager layout) {
            int dragFlags = ItemTouchHelper.UP | ItemTouchHelper.DOWN;

            final LinearLayoutManager linearLayout = (LinearLayoutManager) layout;
            if (linearLayout.getOrientation() == LinearLayoutManager.HORIZONTAL) {
                dragFlags = ItemTouchHelper.LEFT | ItemTouchHelper.RIGHT;
            }
            return dragFlags;
        }

        @Override
        int getSwipeFlags(final RecyclerView.LayoutManager layout) {
            int swipeFlags = ItemTouchHelper.RIGHT;

            final LinearLayoutManager linearLayout = (LinearLayoutManager) layout;
            if (linearLayout.getOrientation() == LinearLayoutManager.HORIZONTAL) {
                swipeFlags = ItemTouchHelper.UP;
            }
            return swipeFlags;
        }
    },
    GRID {
        @Override
        int getDragFlags(final RecyclerView.LayoutManager layout) {
            return ItemTouchHelper.UP | ItemTouchHelper.DOWN | ItemTouchHelper.LEFT | ItemTouchHelper.RIGHT;
        }

        @Override
        int getSwipeFlags(final RecyclerView.LayoutManager layout) {
            int swipeFlags = ItemTouchHelper.RIGHT;

            final GridLayoutManager gridLayout = (GridLayoutManager) layout;
            if (gridLayout.getOrientation() == LinearLayoutManager.HORIZONTAL) {
                swipeFlags = ItemTouchHelper.UP | ItemTouchHelper.DOWN;
            }
            return swipeFlags;
        }
    },
    STAGGERED {
        @Override
        int getDragFlags(final RecyclerView.LayoutManager layout) {
            return ItemTouchHelper.UP | ItemTouchHelper.DOWN | ItemTouchHelper.LEFT | ItemTouchHelper.RIGHT;
        }

        @Override
        int getSwipeFlags(final RecyclerView.LayoutManager layout) {
            int swipeFlags = ItemTouchHelper.RIGHT;

            final StaggeredGridLayoutManager staggeredGridLayout = (StaggeredGridLayoutManager) layout;
            if (staggeredGridLayout.getOrientation() == StaggeredGridLayoutManager.HORIZONTAL) {
                swipeFlags = ItemTouchHelper.UP | ItemTouchHelper.DOWN;
            }
            return swipeFlags;
        }
    };

    /**
     * Returns drag flags for the given layout manager.
     * @param layout layout manager instance
     * @return drag flags
     */
    abstract int getDragFlags(final RecyclerView.LayoutManager layout);

    /**
     * Returns swipe flags for the given layout manager.
     * @param layout layout manager instance
     * @return swipe flags
     */
    abstract int getSwipeFlags(final RecyclerView.LayoutManager layout);
}

File: library/src/main/java/com/thesurix/gesturerecycler/LayoutFlags.kt
package com.thesurix.gesturerecycler


import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.StaggeredGridLayoutManager
import android.support.v7.widget.helper.ItemTouchHelper

/**
 * Enum with predefined gesture flags for various layout managers, see [RecyclerView.LayoutManager]
 * @author thesurix
 */
internal enum class LayoutFlags {
    LINEAR {
        override fun getDragFlags(layout: RecyclerView.LayoutManager): Int {
            val linearLayout = layout as LinearLayoutManager
            return when(linearLayout.orientation) {
                LinearLayoutManager.HORIZONTAL -> ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT
                else -> ItemTouchHelper.UP or ItemTouchHelper.DOWN
            }
        }

        override fun getSwipeFlags(layout: RecyclerView.LayoutManager): Int {
            val linearLayout = layout as LinearLayoutManager
            return when(linearLayout.orientation) {
                LinearLayoutManager.HORIZONTAL -> ItemTouchHelper.UP
                else -> ItemTouchHelper.RIGHT
            }
        }
    },
    GRID {
        override fun getDragFlags(layout: RecyclerView.LayoutManager): Int {
            return ItemTouchHelper.UP or ItemTouchHelper.DOWN or ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT
        }

        override fun getSwipeFlags(layout: RecyclerView.LayoutManager): Int {
            val gridLayout = layout as GridLayoutManager
            return when(gridLayout.orientation) {
                GridLayoutManager.HORIZONTAL -> ItemTouchHelper.UP or ItemTouchHelper.DOWN
                else -> ItemTouchHelper.RIGHT
            }
        }
    },
    STAGGERED {
        override fun getDragFlags(layout: RecyclerView.LayoutManager): Int {
            return ItemTouchHelper.UP or ItemTouchHelper.DOWN or ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT
        }

        override fun getSwipeFlags(layout: RecyclerView.LayoutManager): Int {
            val staggeredGridLayout = layout as StaggeredGridLayoutManager
            return when(staggeredGridLayout.orientation) {
                StaggeredGridLayoutManager.HORIZONTAL -> ItemTouchHelper.UP or ItemTouchHelper.DOWN
                else -> ItemTouchHelper.RIGHT
            }
        }
    };

    /**
     * Returns drag flags for the given layout manager.
     * @param layout layout manager instance
     * @return drag flags
     */
    internal abstract fun getDragFlags(layout: RecyclerView.LayoutManager): Int

    /**
     * Returns swipe flags for the given layout manager.
     * @param layout layout manager instance
     * @return swipe flags
     */
    internal abstract fun getSwipeFlags(layout: RecyclerView.LayoutManager): Int
}
